### PR TITLE
Added types for steamapi

### PR DESCRIPTION
**Pull request by joshuajeschek**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/steamapi/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
+// Type definitions for steamapi 2.1
+// Project: https://github.com/xDimGG/node-steamapi
+// Definitions by: vanitasboi <https://github.com/vanitasboi>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+declare class SteamAPI {
+        constructor(key: string, options?: ConstructorOptions);
```
Comment (peterblazejewicz) [inline types/steamapi/index.d.ts:7]:
```suggestion
        constructor(key: string, options?: Options);
```
Just `Options` would align better with api and readme

File: types/steamapi/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        getUserGroups(id: string): Promise<string[]>;
+        getUserLevel(id: string): Promise<number>;
+        getUserOwnedGames(id: string): Promise<Game[]>;
+        getUserRecentGames(id: string): Promise<RecentGame[]>;
+        getUserServers(hide?: boolean, key?: string): Promise<PlayerServers>;
+        getUserStats(id: string, app: string): Promise<PlayerStats>;
+        getUserSummary(id: string): Promise<PlayerSummary>;
+}
+
+interface ConstructorOptions {
```
Comment (peterblazejewicz) [inline types/steamapi/index.d.ts:32]:
```suggestion
interface Options {
```

File: types/steamapi/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+declare class SteamAPI {
+        constructor(key: string, options?: ConstructorOptions);
+        get(path: string, base?: string, key?: string): object;
+        resolve(info: string): Promise<string>;
+        getAppList(): Promise<App[]>;
+        getFeaturedCategories(): Promise<object[]>;
+        getFeaturedGames(): Promise<object>;
+        getGameAchievements(app: string): Promise<object>;
+        getGameDetails(app: string, force?: boolean): Promise<object>;
```
Comment (peterblazejewicz) [inline types/steamapi/index.d.ts:14]: does it require optional `region` param?
https://github.com/xDimGG/node-steamapi/blob/9b973aaf354daaaaef17dcb35a2ea641460379d5/src/SteamAPI.js#L148

Comment (vanitasboi) [reply types/steamapi/index.d.ts:14]: yes, you are right. that would be an optional string then

Comment (vanitasboi) [reply types/steamapi/index.d.ts:14]:
```suggestion
        getGameDetails(app: string, force?: boolean, region?: string): Promise<object>;
```

**Comment by peterblazejewicz** [comment]
@vanitasboi you can consider changing generic `object` to something like `Record<string, unknown>` to offer more flexiblity
Also you can consider moving your other classes to proper `src/structures` directory, as the module allows to import them directly:
```ts
import Server = require('steamapi/src/structures/Server')
```
At the moment, if you stick with current implementation, you can switch to interfaces instead of classes, as those are not exported and cannot be created. but when you switch to proper `src/structures` and just import to `index.d.ts`, this would improve types interopability. Let me knwo what you think

**Comment by vanitasboi** [comment]
@peterblazejewicz Sounds great! Did not know I could do the importing like this :)
Gonna try this and push again...

**Comment by vanitasboi** [comment]
@peterblazejewicz I switched to Records, but the thing with `src/structures`, I do not understand.
When trying the snippet you sent, the  compiler complained that there where no declaration files for the module 'steamapi/src/structures/Server' and there is also a rule against global imports of itself.
Could you elaborate on this way to import Classes?

**Comment by RyanCavanaugh** [comment]
@peterblazejewicz I didn't see deep importing as a documented use case when skimming the library; is this encouraged?

**Comment by peterblazejewicz** [comment]
@RyanCavanaugh no, you're correct. That's not documented, assumed not intended usage. It works at runtime, but I don't expect people using this. Same for other wrappers I've seen, usually wrapper for api is created exposing common methods.

@vanitasboi no worries. Yes, to make it work, you would have to restructure definition properly. Stick with current module. As pointed, it works just fine for expected usage, thx!

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@vanitasboi thx!

**Comment by vanitasboi** [comment]
Ready to merge ^^

**Pull request merged**